### ISSUE TITLE
Backward regex search stops at the first Non-ASCII character, in Unicode encoded files

### ISSUE THREAD

**Issue by guy038** [notepad-plus-plus/notepad-plus-plus#9553]
Dear **Contributers** and **All**,

We suppose that this attribute : **`regexBackward4PowerUser="yes"`**, in the **`<FindHistory •••••• <FindHistory/>`** tag, of the **config.wml** file, in order to allow **backward** regexes search, is **set** !

### Description of the Issue :

The simple regex **`(?-s).`** **stops** searching when it meets **any** character with code-point **`> 127`**, in an **`Unicode`** encoded file !

### Steps to Reproduce the Issue :

- Paste the text, below, in a N++ **new** tab :

~~~z
Line 1 This isa a test
Line 2 This is~ a test
Line 3 
Line 4 This is€ a test
Line 5 This isƒ a test
Line 6 This isÀ a test
Line 7 This isÿ a test
~~~

- If necessary, use the **`Encoding > Convert to UTF-8`** option to get an **`Unicode`** encoded file

- Open the **Find** dialog ( **`Ctrl + F`** )

    - SEARCH **`(?-s).`**

    - Tick the **`Wrap around`** option
	
	- Select the **`Regular expression`** search mode
	
    - Click, **repeatedly**, on the **`Find Next`** or the **`F3`** key
	
=> As expected, **one standard** character is **selected**, at a time, on subsequent lines !
	
- Now, search in **backward** direction with **`Shift + Enter`** or **`Shift + F3`**
	
=> This time, as soon you reach the character **right after** the string **"is"**, in lines **`4`** to **`7`**( characters with code-point **over** **`007F`** ), the search **stops** and cannot match the letter **`s`** of the word "**is**", located **before** :-((. This issue occurs also for instance, with the **`\w`** regex.

**Notes** : 

- This issue does **not** occur when the current file is **`ANSI`** encoded !

- This issue exists, unfortunately, in **all** **`Unicode`** N++ releases :-(

### Expected Behavior :

The **backward** regex search should match the **same** characters than when performing a **forward** regex search !

### Debug-Info :

~~~z
Notepad++ v7.8.5   (32-bit)
Build time : Mar  4 2020 - 11:04:20
Path : D:\@@\785\notepad++.exe
Admin mode : OFF
Local Conf mode : ON
OS Name : Microsoft Windows XP (32-bit) 
OS Build : 2600.0
Plugins : BetterMultiSelection.dll ChangedLines.dll ComparePlus.dll DSpellCheck.dll ElasticTabstops.dll Explorer.dll mimeTools.dll NppConverter.dll NppExport.dll NppPluginDemo.dll NppQCP.dll NppToolBucket.dll PythonScript.dll SurroundSelection.dll 
~~~

Best Regards,

guy038

**Comment by sasumner** [comment]
@guy038 

Are you really reporting this on 7.8.5 ??

Are you really reporting this on the XP operating system ??  :-)  :-)

Are you really reporting a backward regex search problem, when we've decided that backward regex search is not supported in Notepad++, and that if you enable the "power user" mode for it, it is "at your own risk" of getting poor results?

If you want to find regex results in a backward fashion, run a "Find All..." search and then examine the result that is indicated before your current position.

**Comment by sasumner** [comment]
Of course, perhaps we apply a similar technique as the still-unproven #9552 in order to find "backward" matches by actually moving forward.  But we need your help in testing 9552 first.

**Comment by sasumner** [comment]
So what happens here is that when the search fails, it is passing to the lower-level search routines a search-end position of zero (correct) and a search-start position that is in the "interior" of a three-byte UTF8 character (in the case of the Line 4 data of `€` ).  This is clearly a bad thing to do, only positions that fall on UTF8 character boundaries should be passed in.

**Comment by guy038** [comment]
Hi, @sasumner,

How are you, **Scott** ? Still away from **Covid-19** ?

Sorry, my **previous** post wasn't **accurate** enough !

As you know, since the **`7.9.3`** release, **`Windows XP`** is **not** supported ! So, I'm presently getting some **documentation**, on Net and shops, about all **recent** computer technologies, before buying a decent **`Win 10`** laptop, in **replacement** of my **`Win XP`** antiquity !!

---

So, this morning, I downloaded the **last** portable N++ **`v7.9.3`** release in my wife's **`Win 8`** laptop ( I tried to **hide** that **`793`** folder icon on her **desktop**, in a place rather hidden by the **background** image;-)) Shhhhh ! )

And I can confirm that this issue **still** occurs, with the **last** release. **Idem**, when running the command **`notepad++.exe -noPlugin`**

---

On the other hand, I **verified** that this issue occurs, **both** :

- From the **`v7.8.7`** and **next** releases ( so with the **`regexBackward4PowerUser`** attribute, in the **`config.xml`** file )

- From **previous** releases to the **`v7.8.6`** one ( so **without** the **`regexBackward4PowerUser`** attribute )

---

Finally, I, **successively**, did tests on releases **`v7.7`**, **`v6.9.2`**, **`v6.8.1`** and **`v6.4.5`** and this issue, as I said in my **previous** post, seems to occur in **all** **`Unicode`** N++ releases :-(

---

Now, I'm going to study this **PR** : https://github.com/notepad-plus-plus/notepad-plus-plus/pull/9552 ( Many Thanks, **Scott** for providing the **instructions**  https://github.com/notepad-plus-plus/notepad-plus-plus/issues/9493#issuecomment-781344297 )

Best Regards,

guy038

**Comment by sasumner** [comment]
@guy038 

>covid19

So far so good...

>...before buying a decent Win 10 laptop, in replacement of my Win XP antiquity !!

It is too bad that you let the "future" overtake you, without being proactive enough about it.  :-)
Now you are "suffering" a bit.  :-)

>this issue, as I said in my previous post, seems to occur in all Unicode N++ releases :-(

Yes, it makes sense that this problem affects (and was so far unnoticed) in every Unicode version of Notepad++, because (a) it has been a slow evolution to true UTF8 support in N++ and (b) more people are using a wider range of characters now -- including me! (remember when I would call myself an "A-to-Z person"?  :-)

**Comment by sasumner** [comment]
>So what happens here is that when the search fails, it is passing to the lower-level search routines a search-end position of zero (correct) and a search-start position that is in the "interior" of a three-byte UTF8 character (in the case of the Line 4 data of € ). This is clearly a bad thing to do, only positions that fall on UTF8 character boundaries should be passed in.

I had some concerns about what happens when a _forward_ search is conducted when the start position is specified as an "interior" position of a multibyte UTF8 character.  I experimented with this situation and found that this did not cause any trouble to the search routines.

It isn't really relevant to this issue, but it could be relevant to a possible solution, so I wanted to document it here for now.
